docs/practices/pull_requests.md

```diff
@@ -0,0 +1,73 @@
+# Pull request workflow
+
+Our team practices should encourage transparency and collaboration, and reviewing one another's work is a great way to ensure that we share context and knowledge, and improve the codebase.
+Our goals are to balance inclusive and transparent team practices with the reality that we are all asynchronous and don't want to pay a large coordination penalty for every change and decision.
+
+**All changes to a JupyterHub repository should happen via Pull Requests** and **each pull request should reference at least one issue**.
```

Comment:
Yeah, I agree. I wouldn't want to imply that we want folks to open an Issue immediately before opening a PR that closes it. PRs _are_ issues, so I find having a discussion in one place for an issue and its solution is often preferable to having one discussion in two places. If I found a problem, I would only open an Issue first if I knew I didn't have time to open a PR instead, and I'd encourage others to do the same. I think that's easier to keep track of for ~everyone unless you are doing specific tracking/reporting based on closed Issues only, which we don't.

If folks want, we could mention the reasons for opening a PR against an already acknowledged issue, and the potential reasons to open an issue first to discuss a change before working on a solution. I mainly don't want anyone at the stage of 'I have my change, time to open a PR' deciding they need to open an _issue_ first for some reason, and I feel like this text is telling folks to they _should_ do that, when the opposite is true.

Reply:
Ah ok - I will remove this then. My goal in this PR is not to propose anything too new.

For what it's worth, I disagree. I think GitHub is already not very accessible to non-developers, and pull requests are the worst place to make decisions if you want to include non-developers.

I also think that doing all of the work in just a PR (for non-trivial changes) conflates two distinct parts of the process that require input from different stakeholders :

1. Agreeing on the problem to solve, agreeing that it is worth solving, and aligning on a general idea of what the solution looks like (I think both of these should be in issues or a discussion forum)
2. Agreeing on an _implementation_ to the problem and potentially clarifying design considerations that come up during implementation (i think this needs to be PRs)

Obviously I don't think that's a strict process for *all* changes, but for anything non-trivial, yes my preference is to always have an issue or forum topic for discussion first.

Reply:
One concern I have is that we're struggling to stay on top of all issues and PRs across our repositories. Long-lasting issues and PRs are frustrating for contributors or users, as is the lack of well defined process in some cases, but that's _partially_ balanced by being an open-source project which engenders some goodwill and understanding of how open-source works.

That goodwill is lost if we ask contributors to follow a certain process but don't follow-up- if I'm interacting with a project with no guidelines it's frustrating but I understand that the voluntary nature of open-source means things can be a bit haphazard sometimes. It's a lot more frustrating to have guidelines which I've followed but which aren't reciprocated due to lack of time or resources.

Reply:
I think this is a good point. I will try and change the wording tomorrow to be more nuanced

Reply:
Big 👍  to not encouraging people to create issues a few seconds before they create a PR that closes it. For the reasons that Min and Simon articulated. I think a great strength of a PR is that you can have all the discussion and the change right there, basically all the context you need to take action (review, merge, etc) is right there.

On the topic of keeping discussions separate (should we even solve this? vs how to solve this?) I think doing it all in the PR discussion, especially if there is code to look at makes sense. In rare cases, for contentious changes it might make sense to move the discussion to an issue to try and separate it from a particular implementation (and its author). Basically a way to grease the social wheels of making progress with reduced emotional attachment to a particular change. But I think this is rare, so we can do it ad-hoc.

Sometimes the result of a PR might be "this is a problem, but somehow this doesn't feel like the right implementation and we don't know what the right implementation looks like/don't have time to work on it" - in these cases I think it makes sense to open an issue with a short summary of what happened in the PR to "reboot" the conversation and make it easier to get into it when revisiting it later.

Reply:
I've re-worked the language in the latest commit to be softer on this statement, can you focus discussion on that language to suggest edits etc? I'm not going to debate when we should/shouldn't open issues (my opinion is that we should always open issues, but I recognize I'm in the minority on that one and I don't feel the need to convince people here as long as the language in this PR is OK to y'all).

Reply:
Interesting! I don't view issues and PRs as distinguishable for conversation purposes in terms of participants or goals, other than PRs having concrete changes to discuss, while Issues don't. Forums/Discussions are definitely a mechanism for broadening participation. I've never had the impression that Issues have different visibility/participation than PRs.

Suggesting opening an Issue for discussion prior to PRs seems sensible. Maybe also suggesting soliciting feedback for large/user-facing changes via wider mediums like the forum even after opening a PR is a good idea, too.

Reply:
My thinking on this changed considerably once I started using GitHub as a collaboration platform with people who were not developers. In my experience, when a non-developer is asked to participate in a discussion, but then sees a page that displays a bunch of source code, they are often significantly discouraged to participate because they feel like they're out of their element.

I also find that putting the code in front of everybody's face tends to focus discussions on implementation-centric ideas, which are often very different from design- and UX-level ideas. Pull Requests are inherently a code-centric thing, and much of the activity around deciding "how should our tools change?" does not require understanding our codebase. In fact, many of the people out there with deep expertise in UX and product design are not coders at all! I think it's important for somebody to be able to be heard without requiring them to look at source code.

Ah I can feel myself slipping into a debate about this 😅 as I say I am happy to just stick with the minimal language that will let us move forward, and I'm happy to have deeper discussions about inclusion and accessibility in our team processes in follow-ups.

Reply:
Those are excellent points, and I don't think I appreciated the _presence_ of code being off-putting for folks otherwise happy with Issues. My main experience has been that Issues are _similarly exclusionary_ to PRs, not that PRs aren't exclusive. So in the goal of soliciting broader participation, I haven't found increasing the use of granular Issues improves things over discussion in PRs to be beneficial, whereas _other_ lower-traffic, non-code-centric places like the forum/projects/discussions/blogs/social media definitely is.

But I agree, we should always be working to improve our processes for inclusion, and that's a bigger, broader topic. Thanks for always being so thoughtful on this!